### ISSUE TITLE
feat: Dot access for data hashes (front matter, YAML/JSON files, etc.)

### ISSUE THREAD

**Issue by jaredcwhite** [bridgetownrb/bridgetown#144]
I've been experimenting with ways to provide dot access (aka `page.data.title`, `site.metadata.twitter.handle`, etc.) within Ruby templates/plugins. By far I find it the most annoying aspect of transitioning from Liquid to ERB/etc.

Since Bridgetown already switched from plain hashes to using `ActiveSupport::HashWithIndifferentAccess`, the next logical step would be for that to provide dot access as well. Oddly, I wasn't able to find any widely-referenced example for how to add that, so this is my best attempt at a monkey-patch:

```ruby
ActiveSupport::HashWithIndifferentAccess.class_eval do
  def respond_to_missing?(key, *)
    return true if "#{key}".end_with?("=")

    key?[key]
  end

  def method_missing(key, *args)
    if "#{key}".end_with?("=")
       self["#{key}".chop] = args.first
    elsif self.key?(key)
      self[key]
    else
      default
    end
  end
end
```

My feeling is that this should be added as a feature out-of-the-box, but if there's some reason it doesn't sit right with anyone (language purity? performance?), the monkey-patch could be disabled with an ENV var or something.

There's also the issue of keys colliding with valid Ruby methods (couldn't have a front matter key called `each` for obvious reasons), but that's no different than with ActiveRecord and database column names, so I think most Ruby programmers will understand the caveats involved, and there's always the typical `[]` access method to fall back on.

**Comment by jaredcwhite** [comment]
Another nice thing about this is doing `data&.something&.something_else&.value`…yes, I know the `dig` method exists for hashes but it's not my favorite pattern TBH.

**Comment by ParamagicDev** [comment]
I have a blog post on how I did this in Snowpacker if you wanna take a look (yes there's some optimizations I could've made, but you get the idea)

https://blog.konnor.site/ruby/dynamic_getters_and_setters/

**Comment by jaredcwhite** [comment]
I gave this some more thought, and while dot access could be reasonably tested not to cause any issues within the Bridgetown codebase, the same couldn't be said for other gems—or future integration points with Rails for that matter. As an alternative, I had experimented with a subclass of `HashWithIndifferentAccess` called `HashWithDotAccess`, and if we switch to that specifically, we don't have to worry about any monkey patch. So I think that's a preferable way to go.
